staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -429,58 +535,252 @@ func TestMuxAndDiscoveryComplete(t *testing.T) {
+				// and then send a series of requests to the apiserver, and
+				// we expect these series of requests to be completed successfully
+				<-s.lifecycleSignals.AfterShutdownDelayDuration.Signaled()
+
+				// we send 5 requests, once every second
+				var err error
+				client := newClient(true)
+				for i := 0; i < 5; i++ {
+					r := doer.Do(client, func(httptrace.GotConnInfo) {}, fmt.Sprintf("/echo?message=attempt-%d", i), 1*time.Second)
+					err = r.err
+					if err == nil && r.response.StatusCode != http.StatusOK {
+						err = fmt.Errorf("did not get status code 200 - %#v", r.response)
+						break
+					}
+					time.Sleep(time.Second)
+				}
+				preShutdownHookErrCh <- err
+				return nil
+			})
+			if err != nil {
+				t.Fatalf("Failed to add pre-shutdown hook - %v", err)
+			}
 
-	preShutdownHookErrCh := make(chan error)
-	err := s.AddPreShutdownHook("test-backend", func() error {
-		// this pre-shutdown hook waits for the requests in flight to drain
-		// and then send a series of requests to the apiserver, and
-		// we expect these series of requests to be completed successfully
-		<-s.lifecycleSignals.InFlightRequestsDrained.Signaled()
-
-		// we send 5 requests, once every second
-		var r result
-		client := newClient(true)
-		for i := 0; i < 5; i++ {
-			r = doer.Do(client, func(httptrace.GotConnInfo) {}, fmt.Sprintf("/echo?message=attempt-%d", i), 1*time.Second)
-			if r.err != nil {
-				break
+			// start the API server
+			stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+			go func() {
+				defer func() {
+					// this test has an inherent race condition when we wait for two go routines
+					// to finish - the Run method and the pre-shutdown hook, each running in
+					// its own goroutine, give it a second before unblocking the test assert
+					<-time.After(time.Second)
+					close(runCompletedCh)
+				}()
+				s.PrepareRun().Run(stopCh)
+			}()
+			waitForAPIServerStarted(t, doer)
+
+			close(stopCh)
+
+			select {
+			case err := <-preShutdownHookErrCh:
+				if err != nil {
+					t.Fatalf("PreSHutdown hook can not access the API server - %v", err)
+				}
+			case <-runCompletedCh:
+				t.Fatalf("API Server exited without running the PreShutdown hooks")
+			case <-time.After(15 * time.Second):
+				t.Fatalf("test timed out after 15 seconds")
 			}
-			time.Sleep(time.Second)
-		}
-		preShutdownHookErrCh <- r.err
-		return nil
+		})
+	}
+}
+
+type signalRecorder struct {
+	lock  sync.Mutex
+	order []string
+}
+
+func (r *signalRecorder) before(s lifecycleSignal) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.order = append(r.order, s.Name())
+}
+
+func (r *signalRecorder) verify(got []string) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	want := r.order
+	if !reflect.DeepEqual(want, got) {
+		return fmt.Errorf("Expected order of termination event signal to match, diff: %s", cmp.Diff(want, got))
+	}
+	return nil
+}
+
+type inFlightRequest struct {
```

Comment:
I'm wondering if some of those refactoring-related changes in tests (e.g. this, signalrecorder, changes in wrapping signals, injection of fakeAudit, etc. seems like could just go in without any changes to the logic and tests).

Reply:
yes, i am happy to open a new PR with test refactors if it makes review easier

Reply:
Yes please - it would help.

Reply:
https://github.com/kubernetes/kubernetes/pull/110132